Remove torch.distributed calls from TorchDistributedTrial properties

nlgranger opened the pull request:
Fixes #3484

## Motivation

Removes all calls to torch.distributed from TorchDistributedTrial properties as they have undesired side-effects: namely they break the synchronization of distributed processes if one forgets to call them simultaneously on all distributed processes.

## Description of the changes
<!-- Describe the changes in this PR. -->
Prior to this PR, reading properties on TorchDistributedTrial triggered a call to torch.distributed.broadcast in order to retrieve the values, this was necessary because they are solely stored on the process with rank 0.
This PR instead broadcasts the values to all processes each time they are updated by calls to suggest_uniform, etc. This is not inconvenient because these functions are already called on all distributed processes anyways and Optuna documentation is clear about this.

nlgranger commented:
Does this need a test? If so I'd like help on what should be tested.

toshihikoyanase commented:
> Does this need a test? 

Since this PR adds new functionality, I guess we need tests.
I'd like to check the following two points.

1. The rank-0 node gets property values without hanging processes.
2. The property values are properly updated after changing the values.

I'm think they can confirmed that by adding a following assertions. What do you think of it?

```
@pytest.mark.filterwarnings("ignore::optuna.exceptions.ExperimentalWarning")
@pytest.mark.parametrize("storage_mode", STORAGE_MODES)
def test_params(storage_mode: str) -> None:
    with StorageSupplier(storage_mode) as storage:
        if dist.get_rank() == 0:
            study = optuna.create_study(storage=storage)
            trial = TorchDistributedTrial(study.ask())
        else:
            trial = TorchDistributedTrial(None)

        trial.suggest_float("f", 0, 1)
        trial.suggest_int("i", 0, 1)
        trial.suggest_categorical("c", ("a", "b", "c"))

        params = trial.params
        assert 0 <= params["f"] <= 1
        assert 0 <= params["i"] <= 1
        assert params["c"] in {"a", "b", "c"}
+
+        if dist.get_rank() == 0:
+            assert 0 <= trial.params["f"] <= 1
+            assert 0 <= params["i"] <= 1
+            assert params["c"] in {"a", "b", "c"}
```

toshihikoyanase commented:
In addition, the change will increase communication to update the property values.
Could you briefly check the difference of processing time of functions with the `@updates_state` annotation?

nlgranger commented:
> 1. The rank-0 node gets property values without hanging processes.
> 2. The property values are properly updated after changing the values.

I have added a new test to check that. It loops over all properties so that it doesn't miss one by accident in the future.
Since it checks for deadlocks, I have added a timeout to the test which in turns requires the pytest-timeout package dependency.



> Could you briefly check the difference of processing time of functions with the `@updates_state` annotation?

I have run the following code on my computer (CPU only experiment with gloo distributed backend)

```py
import os
import tempfile
import time

from torch import distributed as dist

import optuna
from optuna.integration import TorchDistributedTrial


def main():
    os.environ["WORLD_SIZE"] = os.environ["OMPI_COMM_WORLD_SIZE"]
    os.environ["RANK"] = os.environ["OMPI_COMM_WORLD_RANK"]
    os.environ["MASTER_ADDR"] = "127.0.0.1"
    os.environ["MASTER_PORT"] = "20000"

    dist.init_process_group("gloo")

    tempdir = tempfile.NamedTemporaryFile()
    url = "sqlite:///{}".format(tempdir.name)
    storage = optuna.storages.RDBStorage(
        url,
        engine_kwargs={"connect_args": {"timeout": 10}}
    )

    if dist.get_rank() == 0:
        study = optuna.create_study(storage=storage)
        trial = TorchDistributedTrial(study.ask())
    else:
        trial = TorchDistributedTrial(None)

    dist.barrier()

    t1 = time.monotonic()

    for i in range(100):
        trial.suggest_float(str(i), 0., 1.)

    dist.barrier()

    t2 = time.monotonic()

    if dist.get_rank() == 0:
        print(t2 - t1)


if __name__ == "__main__":
    main()
```

The timing values are fairly consistent and the update_states causes an an overhead of 23% (0.66s -> 0.83s). Although significant, this is absolutely negligible in a typical PyTorch workflow.

nlgranger commented:
I'm gonna need help with mypy checks, I'm no expert in python typing hints.

toshihikoyanase commented on optuna/integration/pytorch_distributed.py:
```diff
@@ -24,6 +25,42 @@
+    """
+
+    @functools.wraps(f)
+    def wrapped(self: Any, *args, **kwargs) -> Any:
+        def state() -> Sequence:
+            assert self._delegate is not None
+            return (
+                self._delegate.number,
+                self._delegate.params,
+                self._delegate.distributions,
+                self._delegate.user_attrs,
+                self._delegate.system_attrs,
+                self._delegate.datetime_start,
+            )
+
+        try:
+            return f(self, *args, **kwargs)
```
Mypy showed an error due to missing type annotation for the arguments.
Not sure, but the optuna code corresponding to this `wrapped` function takes `args` and `kwargs`, so I guess we can refactor this function as follows:

```suggestion
    def wrapped(*args: Any, **kwargs: Any) -> Any:
        def state() -> Sequence:
            self: TorchDistributedTrial = args[0]
            assert self._delegate is not None
            return (
                self._delegate.number,
                self._delegate.params,
                self._delegate.distributions,
                self._delegate.user_attrs,
                self._delegate.system_attrs,
                self._delegate.datetime_start,
            )

        try:
            return f(*args, **kwargs)
```

In addition, I guess we can use `properties` instead of `state`. What do you think of it?

toshihikoyanase commented on optuna/integration/pytorch_distributed.py:
```diff
@@ -24,6 +25,42 @@
 
 
+def updates_state(f: Callable[..., Any]) -> Callable[..., Any]:
+    """Method decorator to fetch updated trial state from rank 0 after f is run.
```
A nit.
```suggestion
    """Method decorator to fetch updated trial properties from rank 0 after ``f`` is run.
```

toshihikoyanase commented on optuna/integration/pytorch_distributed.py:
```diff
@@ -24,6 +25,42 @@
+
+    This decorator ensures trial properties (params, distributions, etc.) on all distributed
+    processes are up-to-date with the wrapped trial stored on rank 0.
+    It should be applied to all Trial methods that update property values.
```
```suggestion
    It should be applied to all :class:`~optuna.integration.TorchDistributedTrial` methods that update property values.
```

toshihikoyanase commented on optuna/integration/pytorch_distributed.py:
```diff
@@ -73,7 +110,16 @@ def __init__(
             assert trial is None, "error message"
         self._delegate = trial
         self._device = device
+        self._params = {}
```
Can we remove this line?
```suggestion
```

toshihikoyanase commented on tests/integration_tests/test_pytorch_distributed.py:
```diff
@@ -334,3 +334,38 @@ def test_distributions(storage_mode: str) -> None:
+
+@pytest.mark.filterwarnings("ignore::optuna.exceptions.ExperimentalWarning")
+@pytest.mark.parametrize("storage_mode", STORAGE_MODES)
+@pytest.mark.timeout(15, method="signal")
```
In my environment, `pytest-timeout` did not terminate the MPI processes as expected.
Could you check if it terminate the processes if you add following lines to `test_updates_state`?

```
       if dist.get_rank() == 0:
            trial.suggest_float("f", 0, 1)
```

toshihikoyanase replied:
The test finished after 30 minutes. The `torch.distributed` has the [timeout functionality](https://pytorch.org/docs/stable/distributed.html#torch.distributed.init_process_group), and we do not need to introduce `pytest-timeout`.

```console
tests/integration_tests/test_pytorch_distributed.py:347: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
optuna/_experimental.py:90: in wrapped_init
    _original_init(self, *args, **kwargs)
optuna/integration/pytorch_distributed.py:116: in __init__
    self._number = self._broadcast(getattr(self._delegate, "number", None))
optuna/integration/pytorch_distributed.py:286: in _broadcast
    dist.broadcast(size_buffer, src=0)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

tensor = tensor([5], dtype=torch.int32), src = 0, group = None, async_op = False

    def broadcast(tensor, src, group=None, async_op=False):
        """
        Broadcasts the tensor to the whole group.
    
        ``tensor`` must have the same number of elements in all processes
        participating in the collective.
    
        Args:
            tensor (Tensor): Data to be sent if ``src`` is the rank of current
                process, and tensor to be used to save received data otherwise.
            src (int): Source rank.
            group (ProcessGroup, optional): The process group to work on. If None,
                the default process group will be used.
            async_op (bool, optional): Whether this op should be an async op
    
        Returns:
            Async work handle, if async_op is set to True.
            None, if not async_op or if not part of the group
    
        """
        _check_single_tensor(tensor, "tensor")
        if _rank_not_in_group(group):
            return
    
        opts = BroadcastOptions()
        opts.rootRank = src
        opts.rootTensor = 0
    
        if group is None or group is GroupMember.WORLD:
            default_pg = _get_default_group()
            work = default_pg.broadcast([tensor], opts)
        else:
            group_src_rank = _get_group_rank(group, src)
            opts.rootRank = group_src_rank
            work = group.broadcast([tensor], opts)
        if async_op:
            return work
        else:
>           work.wait()
E           RuntimeError: Application timeout caused pair closure

.venv/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py:1167: RuntimeError
```

nlgranger replied:
Fixed by simply lowering the distributed process group default timeout.

toshihikoyanase commented on tests/integration_tests/test_pytorch_distributed.py:
```diff
@@ -334,3 +334,38 @@ def test_distributions(storage_mode: str) -> None:
+
+        # rank 0 can read properties without messing up distributed state
+        if dist.get_rank() == 0:
+            [getattr(trial, p) for p in property_names]
```
[Note] This test does not confirms that the values of properties are correct. But it is reasonable since the value of properties are checked in the other test cases such as `test_params`.

toshihikoyanase commented on tests/integration_tests/test_pytorch_distributed.py:
```diff
@@ -334,3 +334,38 @@ def test_distributions(storage_mode: str) -> None:
+            if isinstance(getattr(TorchDistributedTrial, p), property)
+        ]
+
+        # rank 0 can read properties without messing up distributed state
```
Cosmetic fix to keep the [coding style](https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#block-and-inline-comments-are-supposed-to-be-sentences). In addition, could you clarify what does the `messing up distributed state` mean? The following is just an example.
```suggestion
        # Rank 0 can read properties without deadlock.
```

nlgranger replied:
I mean the state of the distributed context remains synchronized between processes. ie: without deadlock indeed.

toshihikoyanase commented on tests/integration_tests/test_pytorch_distributed.py:
```diff
@@ -334,3 +334,38 @@ def test_distributions(storage_mode: str) -> None:
+
+        dist.barrier()
+
+        # same with rank 1
```
ditto. Just a cosmetic fix.
```suggestion
        # Same with rank 1.
```

toshihikoyanase reviewed (commented):
Thank you for your swift action!

> I'm gonna need help with mypy checks, I'm no expert in python typing hints.

I add suggestion to fix type hints.

> The timing values are fairly consistent and the update_states causes an an overhead of 23% (0.66s -> 0.83s). Although significant, this is absolutely negligible in a typical PyTorch workflow.

Thank you for sharing the speed benchmark. This change will add 1.7 msec per method call, and I think it is acceptable as you mentioned.

> Since it checks for deadlocks, I have added a timeout to the test which in turns requires the pytest-timeout package dependency.

`pytest-timeout` may not work with mpi jobs. Please check my comments below.

nlgranger commented:
Hi. Is there anything left to resolve?

toshihikoyanase commented on optuna/integration/pytorch_distributed.py:
```diff
@@ -24,6 +25,45 @@
 )
 
 
+def updates_properties(f: Callable[..., Any]) -> Callable[..., Any]:
```
@himkt mentioned IDEs cannot infer argument types of the annotated methods. Could you use `ParamSpec` as can be seen in https://github.com/optuna/optuna/pull/3270?

toshihikoyanase reviewed (commented):
@nlgranger Thank you for your kind ping. I just came back from vacation.

@HideakiImamura The change almost looks great to me except for the type inference. Could you assign another reviewer, please?

nlgranger commented:
The mypy test will pass in the next release (https://github.com/python/mypy/pull/12668), I think I'd rather wait than find a workaround if that's ok for you.

toshihikoyanase commented on optuna/integration/pytorch_distributed.py:
```diff
@@ -19,11 +23,54 @@
+
+    @functools.wraps(f)
+    def wrapped(*args: _P.args, **kwargs: _P.kwargs) -> _T:
+        self: TorchDistributedTrial = args[0]
```
Thank you for your update. Could you ignore the type checking tentitatively?
```suggestion
        # TODO(nlgranger): Remove type ignore after mypy includes
        # https://github.com/python/mypy/pull/12668
        self: TorchDistributedTrial = args[0]  # type: ignore
```

toshihikoyanase reviewed (approved):
Thanks for your update!
The suggestion successfully worked in my VS Code.
![image](https://user-images.githubusercontent.com/3255979/169264544-a4536d76-5e91-4dac-857c-a28ef8a8e9a3.png)

I'm sorry but it took some time to notice that `@experimental` prevents the type inference as well.

LGTM!

@contramundum53 Please merge after your review.

nzw0301 commented:
I'll review this PR until the end of next week.

nzw0301 commented on optuna/integration/pytorch_distributed.py:
```diff
@@ -91,20 +148,24 @@ def func() -> float:
         return self._call_and_communicate(func, torch.float)
 
     @deprecated("3.0.0", "6.0.0", text=_suggest_deprecated_msg)
+    @updates_properties
```
Since this method calls `suggest_float` that is decorated by `updates_properties`, can we remove `updates_properties`?
```suggestion
```

I suppose the other deprecated `suggest_*` methods too.

nlgranger replied:
Yes it is possible but I though if all methods are decorated one can change the implementation without having to remember about adding it.

nzw0301 replied:
Thanks. Hmm, but in my understanding, when we call one of these deprecated suggest methods, `updates_properties` is executed twice, which is supposed to be unnecessary communications among processes. 

@toshihikoyanase  What do you think about this?

toshihikoyanase replied:
Good catch! `suggest_uniform` simply calls `suggest_float` and `suggest_float` actually update the properties. So, @nzw0301  suggestion sounds reasonable to me.

> one can change the implementation without having to remember about adding it.

That makes sense, but we will not update the deprecated methods so frequently.

nlgranger replied:
I have removed the redundant decorators in the latest commit.

nzw0301 reviewed (changes requested):
I locally confirmed that the properties were shared correctly among nodes by using `pytest`.  Basically, LGTM. 

I have one comment. Could you take a look at my comment?

nzw0301 commented on optuna/integration/pytorch_distributed.py:
```diff
@@ -19,11 +23,56 @@
+        # https://github.com/python/mypy/pull/12668
+        self: TorchDistributedTrial = args[0]  # type: ignore
+
+        def state() -> Sequence:
```
How about replacing `state` with `fetch_properties` or something else because `state` sounds a little bit vague? 
```suggestion
        def fetch_properties() -> Sequence:
```

nlgranger replied:
I updated the decorator name to match as well.

nzw0301 replied:
thanks!

nzw0301 reviewed (changes requested):
Thanks. I have one more comment.

nzw0301 reviewed (approved):
LGTM. Thanks!

nlgranger commented:
Thank you all for the review work!
